storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoArray.java

```diff
@@ -286,6 +286,34 @@ public void markNodeInvalid(final Bytes32 blockRoot, final Optional<Bytes32> lat
     applyDeltas(new LongArrayList(Collections.nCopies(getTotalTrackedNodeCount(), 0L)));
   }
 
+  public void findAndMarkInvalidChain(final Bytes32 headRoot, final Bytes32 latestValidHash) {
+    final Optional<Integer> maybeIndex = indices.get(headRoot);
+    if (maybeIndex.isEmpty()) {
+      LOG.debug("Couldn't find chain head {} because it was unknown", headRoot);
+      return;
+    }
+    final ProtoNode headNode = getNodeByIndex(maybeIndex.get());
+    if (headNode.getExecutionBlockHash().equals(latestValidHash)) {
+      return;
+    }
+
+    final Optional<Integer> maybeFirstInvalidNodeIndex =
+        findFirstInvalidNodeIndex(maybeIndex.get(), latestValidHash);
+    // In case we have not found latestValidHash, end this up
+    if (maybeFirstInvalidNodeIndex.isEmpty()) {
+      return;
+    }
+
+    final int index = maybeFirstInvalidNodeIndex.get();
+    final ProtoNode node = getNodeByIndex(index);
+    markNodeValid(node.getParentRoot());
+    node.setValidationStatus(INVALID);
+    removeBlockRoot(node.getBlockRoot());
+    markDescendantsAsInvalid(index);
+    // Applying zero deltas causes the newly marked INVALID nodes to have their weight set to 0
+    applyDeltas(new LongArrayList(Collections.nCopies(getTotalTrackedNodeCount(), 0L)));
```

Comment:
It feels like we're duplicating a lot of code here that we also have in `markNodeInvalid` when the only difference should be that in the existing `markNodeInvalid` method we would always mark the `headRoot` as invalid and we would skip that in this version. Given the complexity involved in this duplicated code it would be good to try and refactor to avoid the duplication.

Reply:
Yeah, maybe some boolean flag, will try to refactor now